system/lib/wasmfs/file.h

```diff
@@ -134,6 +146,20 @@ class Directory : public File {
 
     void setEntry(std::string pathName, std::shared_ptr<File> inserted) {
       getDir()->entries[pathName] = inserted;
+      // Simulataneously, set the parent of the inserted node to be this Dir.
+      // inserted must be locked because we have to go through Handle.
+      inserted->locked().setParent(file);
```

Comment:
I would guess this lock should be over both the operations done in this function.

Otherwise, imagine that two directories are racing here, each trying to assign the file to them. They could both do the first operation in this function, that is, add a reference to the file. And then locking would ensure an order between them on the second operation, but one would trample the other's value. So we'd end up with a file referred to by two directories.

Or is that impossible for some other reason, like we have a guarantee of holding some other lock before calling this function?

Reply:
I agree, here it makes sense to acquire the lock over both operations.

Reply:
Meanwhile, I'm not sure even locking the file over both operations is enough. If we try to attach the same file to two separate dirs, we'd end up with the later operation succeeding entirely, but it still tramples half the first operation.

That is, the first thread would set

```
dir1-> file
file-> dir1
```

and then the second thread would set

```
dir2-> file
file-> dir2
```

The final result is

```
dir1-> file
dir2-> file
file-> dir2
```

So we still have two dirs that refer to this file. To fix this, I think we to do a third operation under the lock: first, if the file is already attached to a dir, to detach it from that other dir. This is tricky though as it does mean we need more than one lock held, so that's probably not the right solution.

The fuzzer should be good at finding bugs like this. We should add a "validation" function that the fuzzer would run, and that would do things like scan the entire FS and verify each file appears once in the tree, and each parent pointer matches what it should, etc.

Reply:
This is an important point. As we discussed in our meeting, we can put an assert for now that the `created` file does not have a parent since it is a new file. However, we talked about the `rename` syscall in which there would be an existing source directory, source file and destination directory. Having some sort of try_lock mechanism that would return `EBUSY` in the event that all 3 resources cannot be atomically locked seems like the right move for now. I'll make a note of this in the code as well to re-visit when we implement `rename`.